admin_frontend/src/session.rs

```diff
@@ -62,68 +72,144 @@ impl SessionStorage {
     tracing::info!("del user session: {} res: {}", session_id, res);
     Ok(())
   }
+
+  pub async fn put_code_session(
+    &self,
+    code: &str,
+    code_session: &CodeSession,
+  ) -> redis::RedisResult<()> {
+    let key = code_session_key(code);
+    self
+      .redis_client
+      .clone()
+      .set_options(
+        key,
+        code_session,
+        redis::SetOptions::default().with_expiration(redis::SetExpiry::EX(60 * 5)), // code is valid for 5 minutes
+      )
+      .await
+  }
 }
 
 #[derive(Debug, Serialize, Deserialize)]
-pub struct UserSession {
+pub struct CodeSession {
   pub session_id: String,
-  pub token: GotrueTokenResponse,
+  pub code_challenge: Option<String>,
+  pub code_challenge_method: Option<String>,
+}
+struct CodeSessionOptional(Option<CodeSession>);
+
+impl ToRedisArgs for CodeSession {
+  fn write_redis_args<W>(&self, out: &mut W)
+  where
+    W: ?Sized + redis::RedisWrite,
+  {
+    let s = serde_json::to_string(self).unwrap();
+    out.write_arg(s.as_bytes());
+  }
 }
 
-impl UserSession {
-  pub fn new(session_id: String, token: GotrueTokenResponse) -> Self {
-    Self { session_id, token }
+impl FromRedisValue for CodeSessionOptional {
+  fn from_redis_value(v: &redis::Value) -> redis::RedisResult<Self> {
+    let bytes = expect_redis_value_data(v)?;
+    match bytes {
+      Some(bytes) => {
+        let session = expect_redis_json_bytes(bytes).unwrap();
```

Comment:
is it safe to unwrap? Or we can return None if `expect_redis_json_bytes` returns error

Reply:
it's safe in the sense that the server is the one serializing and de serializing, and only will panic if there is some corruption. but nonetheless, I'll use the `?` to return the error